DIPs/7NNN-WGB.md

```diff
@@ -0,0 +1,164 @@
+
+## Abstract
+
+Required.
+
+Short and concise description of the idea in a few lines.
+
+
+## Contents
+* [Rationale](#rationale)
+* [Prior Work](#prior-work)
+* [Description](#description)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Reference](#reference)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+
+1. Although this supports reordering, the real reason for naming is so one can
+have
+a function with a longish list of parameters, each with a reasonable default,
+and the user need only supply the arguments that matter for his use case. This
+is much more flexible than the current method of putting all the defaults at
+the
+end of the parameter list, and defaulting one means all the rest get defaulted.
+2. For a longish list of parameters, when the user
+finds himself counting parameters to ensure they line up, then named parameters
+can be most helpful.
+3. A step towards making it possible to replace the brace struct initialization syntax with
+a function-call-like construct. I.e. the initialization of structs is unified with
+struct literals and struct constructors.
+
+## Prior Work
+
+* DIP 1019
+* DIP 1020
+
+Both DIP 1019 and DIP 1020 detail prior work in other languages, which will not be
+repeated here.
+
+
+## Description
+
+This based on the recognition that D already has
+named parameters for [struct initializers](https://dlang.org/spec/struct.html#static_struct_init).
+Closely related are [array initializers](https://dlang.org/spec/arrays.html#static-init-static)
+and [associative array literals](https://dlang.org/spec/hash-map.html#static_initialization).
+
+The addition to the syntax is replacing the definition of `ArgumentList` for function call
+arguments with:
+
+```
+ArgumentList:
+   NamedArgument
+   NamedArgument ,
+   NamedArgument , NamedArgumentList
+
+NamedArgument:
+   Identifier : Argument
+   Argument
+```
+
+Assignment of arguments to parameters would work the same way as
+[assignments to fields](https://dlang.org/spec/struct.html#static_struct_init).
+
+Matching of the `NamedArgument`s to a function declaration proceeds in lexical order.
+For each `NamedArgument`:
+
+1. If an `Identifier` is present, it matches to the `Parameter` with the corresponding
+`Identifier`. If it does match any named `Parameter`, then it is an error.
+2. If an `Identifier` is not present, the `Parameter` matched is the one following
+the previous matched `Parameter`, or the first `Parameter` if none have been matched yet.
+3. Matching a `Parameter` more than once is an error.
+4. After the `NamedArgumentList` is exhausted, any unmatched `Parameter`s receive the
+corresponding default value specified for that `Parameter`. If there is no default value,
+it is an error.
+5. If there are more `NamedArgument`s than `Parameter`s, the remainder match the trailing
+`...` of variadic parameter lists, and `Identifier`s are not allowed.
```

Comment:
I wouldn't want to hold this up with my request, but I just wanna through out one idea that might be useful to me in some case: instead of banning identifiers on a variadic, actually use them to make an anonymous struct.

For example, given:

```
void foo(T...)(T args) {}

foo(a: 0, b: "zero");
```

Instead of issuing an error, that actually generates something like this:

```
struct _anonymous {
   int a;
   string b;
}

foo(_anonymous(0, "zero"));
```

and thus `args[0]` is actually an instance of this generated type, holding all the named parameters that matched the variadic.

A practical application of this would be library tuples with named members, or syntax sugar for creating dynamic (e.g. json) objects,or some cases of like database query DSLs.

I don't actually wanna sidetrack y'all too much, just throwing out this as something to consider as potential in changing this rule (or something).

Reply:
So basically Python's `kwargs`?

I'm not against the idea in principle but I think it needs to be opt-in for the callee, otherwise, if a parameter is removed, the error messages are going to be very confusing.

Reply:
yeah, basically. I am inclined to agree on being opt-in too.

Reply:
I like the feature. Another option would be named tuples `foo((a: 0, b: "zero"))`.

Reply:
I agree with @CyberShadow, the DIP as-is provides zero burden on the implementor, they don't have to change anything in any functions, it's just how you can call them is more pleasant.

The sort of change you are proposing would open up existing implementations to a lot of problems (and probably is a breaking change). I get the point of it, but probably not doable. Note that this would also cause problems with template instantiation, since separate calls for the same types would now be different instantiations, depending on the call.

However, *template parameters* might provide a mechanism for this. For example, you already have code like this:

```D
template IKnowYourName(T...)
{
   pragma(msg, __traits(identifier, T[0]));
}

int x;
alias n = IKnowYourName!x; // prints "x"
alias n2 = IKnowYourName!(blah: 0); // could also work, even with existing code
```